### PR TITLE
[14.0][MIG] agreement_legal (Version 12.0 to 14.0)

### PR DESCRIPTION
**Pull request by patrickrwilson**

### CODE REVIEW

**Comment by patrickrwilson** [comment]
@osi-scampbell @ReedHayashikawa can you assist with review?

**Comment by kittiu** [comment]
@patrickrwilson this is the same as https://github.com/OCA/contract/pull/623 ?

cc @newtratip

**Comment by patrickrwilson** [comment]
@kittiu Yes, it's the same. I didn't do a good job checking before submitting this so I commented on @newtratip's PR offering for him to use the tests in this PR to finish making theirs green but haven't received a response yet.

**Comment by kittiu** [comment]
@newtratip since this one already green and it seem to be a complete migration. I think it is easier to go with this PR.

Also, as I checked your PR #623, it has enhancement, i..e, create agreement from template (wizard). So, lets push PR to merge first, and then you can add another to propose the enhancement on top of this one.

WDYT?

File: agreement_legal/__manifest__.py
Comment (kittiu) [inline agreement_legal/__manifest__.py]:
```suggestion
    "version": "14.0.1.0.0",
```

File: agreement_legal/data/demo.xml
Comment (kittiu) [inline agreement_legal/data/demo.xml]:
```suggestion
```
Just the spacing corrections (there are more in this file, I wont' mentioned all).

File: agreement_legal/data/demo.xml
Comment (kittiu) [inline agreement_legal/data/demo.xml:26]:
```suggestion

    <function name="write" model="ir.model.data">
```

File: agreement_legal/models/agreement_appendix.py
Comment (kittiu) [inline agreement_legal/models/agreement_appendix.py]:
```suggestion
        help="The title is displayed on the PDF. The name is not.",
```

File: agreement_legal/models/agreement_clause.py
Comment (kittiu) [inline agreement_legal/models/agreement_clause.py]:
```suggestion
        string="Title", help="The title is displayed on the PDF. The name is not."
```

File: agreement_legal/models/agreement_recital.py
Comment (kittiu) [inline agreement_legal/models/agreement_recital.py:15]:
```suggestion
        help="The title is displayed on the PDF. The name is not.",
```

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py:177]:
```suggestion
        default=lambda self: self._get_default_parties(),
```
https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#38fields

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py:372]:
```suggestion
            if not rec.state == "draft":
                # Make sure status is draft
                rec.state = "draft"
```
I am not sure why we need to reset state of rec to "draft".
Or why not just put {"state": "draft"} in to default_vals.

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py]: Shoulde be "New" or "NEW" (I saw "New" most of the time in odoo code :)

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py:395]: If no for loop, should we have `self.ensure_one()`

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py]:
```suggestion
        return super().create(vals)
```

File: agreement_legal/models/agreement.py
Comment (kittiu) [inline agreement_legal/models/agreement.py]:
```suggestion
        return super().write(vals)
```

File: agreement_legal/static/src/xml/domain_widget_view.xml
Comment (kittiu) [inline agreement_legal/static/src/xml/domain_widget_view.xml:5]: Should we remove this commented section?

**Review by kittiu — APPROVED** [review]

File: agreement_legal/static/src/xml/domain_widget_view.xml
Comment (ygol) [reply agreement_legal/static/src/xml/domain_widget_view.xml:5]: hello, shouldn't "agreement_legal/data/demo.xml" only be loaded if demo data is set?

File: agreement_legal/data/demo.xml
Comment (ygol) [reply agreement_legal/data/demo.xml:26]: hello, shouldn't "agreement_legal/data/demo.xml" only be loaded if demo data is set?

**Comment by patrickrwilson** [comment]
@kittiu Thank you for your review, your suggestions have been implemented.

**Comment by kittiu** [comment]
> @kittiu Thank you for your review, your suggestions have been implemented.

Thanks. Please check travis.

**Comment by patrickrwilson** [comment]
@ygol it's ready for your review.

**Comment by ygol** [comment]
> @ygol it's ready for your review.

I've functionally tested. Those are my findings:

- revision doesn't seem to work. (doesn't list previous versions)

- new version of template loose the "is_template" boolean

- Problem with reference when duplicating an agreement (also when creating from a template).
cf. https://github.com/OCA/contract/issues/640

- problem with "created on" when generating a new version. the archived version get the current date and the new version keeps the origin date (should be switched).
cf. https://github.com/OCA/contract/issues/502

- when adding a child agreement from parent there is no way to save the created child AG + fields should be pre-populated with values from the parent ("parent agreement", "start date", "end date", contacts, etc.)
cf. https://github.com/OCA/contract/issues/500

- I think "data/demo.xml" should only be loaded when demo data is required  https://github.com/OCA/contract/pull/669/files#diff-289fece69d3e767eb82c1fa91fff8a6ae1b74aa76123ef49d0c1aaf49532c9a0R21

**Comment by patrickrwilson** [comment]
@ygol  my last commit has most of your fixes, see my notes below....

* revision doesn't seem to work. (doesn't list previous versions) - FIXED: I added a new field called 'is_old_version' which it now uses to show old versions. Before it was marking them as inactive which is why they weren't visible.
* new version of template loose the "is_template" boolean - FIXED: The agreement module this inherits has copy=False for the is_template fields so in agreement_legal I set it to True
* Problem with reference when duplicating an agreement (also when creating from a template). #640 - This is part of the Agreement module this inherits so I think any changes need to be done in that module.
* problem with "created on" when generating a new version. the archived version get the current date and the new version keeps the origin date (should be switched). #502 - FIXED: I've made it to retain the original create date/time.
* when adding a child agreement from parent there is no way to save the created child AG + fields should be pre-populated with values from the parent ("parent agreement", "start date", "end date", contacts, etc.) #500 - FIXED: I've added context to the one2many field.
* I think "data/demo.xml" should only be loaded when demo data is required  https://github.com/OCA/contract/pull/669/files#diff-289fece69d3e767eb82c1fa91fff8a6ae1b74aa76123ef49d0c1aaf49532c9a0R21 - FIXED: I've removed the items in data/demo.xml since they are already in demo/demo.xml

**Comment by patrickrwilson** [comment]
It looks like I have to fix a test that now fails after these changes. I will work on that shortly.

**Comment by ygol** [comment]
@patrickrwilson thanks so much. here updated remarks. Are those opportune or is it better to first do the migration and leave those fixing for a later PR?

> * revision doesn't seem to work. (doesn't list previous versions) - FIXED: I added a new field called 'is_old_version' which it now uses to show old versions. Before it was marking them as inactive which is why they weren't visible.

Wouldn't be possible to use "context={'active_test': False}" to show those archived revisions? I think there is no way to list "old_version" beside in the history tab now.

> * new version of template loose the "is_template" boolean - FIXED: The agreement module this inherits has copy=False for the is_template fields so in agreement_legal I set it to True

Now when you click on "new agreement" button it creates a template. so if a tradeoff must be done; IMO better to have the original templates loosing their "is_template" when copying. 


> * problem with "created on" when generating a new version. the archived version get the current date and the new version keeps the origin date (should be switched). #502 - FIXED: I've made it to retain the original create date/time.

IMO the date of a version should be the date that version has been created. so if original version was created on 4.5.21 and a new version created on 6.8.21 then version on date would show 4.5.21 and version 2 would show 6.8.21. 
Any other opinion on this?

**Comment by patrickrwilson** [comment]
@ygol I agree that these changes should be in a later PR. The migration should probably migrate it as it was and then list of improvements/fixes made in additional PR's, that way they can properly be vetted and not grouped into the migration task. I will revert the changes back and then we can perform the changes after the migration is merged.

**Comment by patrickrwilson** [comment]
@ygol reverted changes and squashed, should be ready for review and merge, then we can look at improving/fixing those issues.

**Review by osi-scampbell — APPROVED** [review]

**Review by ygol — APPROVED** [review]

**Comment by ygol** [comment]
There's a problem with the report -:(
1. go to an agreement
2. click on preview or print
3. unwanted characters appears in the report

image

**Comment by ygol** [comment]
> There's a problem with the report -:(
> 
> 1. go to an agreement
> 2. click on preview or print
> 3. unwanted characters appears in the report
> 
I think that v14 uses now 'mail_render_mixin' and following 
https://github.com/odoo/odoo/blob/70a75fa1b41f06c10abe02efbb0b4565b4190e45/addons/mail/models/mail_render_mixin.py#L388
it seem a dict is returned. So for example; adding the following line 299 in agreement.py
```
    def _compute_dynamic_description(self):
        MailTemplates = self.env["mail.template"]
        for agreement in self:
            lang = agreement.partner_id.lang or "en_US"
            description = MailTemplates.with_context(lang=lang)._render_template(
                agreement.description, "agreement", [agreement.id]
            )
            des = ""
            for i in description:
                des += description[i]
            agreement.dynamic_description = des
```
seems to fix it.
There's an easier way maybe?

**Comment by patrickrwilson** [comment]
@ygol I've used your suggested code which worked. I'm not sure there is an easier way but this fixed the issue which I'm good with. Hopefully it's good now to merge.

**Comment by hkapatel-initos** [comment]
LGTM Can we please merge this PR?

**Comment by newtratip** [comment]
@patrickrwilson, CI is red, Can you check please ?

**Comment by newtratip** [comment]
@OCA/project-service-maintainers, I think this PR should be merge. I have a lot of enhancement according https://github.com/OCA/contract/pull/623. I will open in another PR.

Thanks!

**Review by newtratip — APPROVED** [review]

**Comment by newtratip** [comment]
@patrickrwilson, I install this module, it's throw error. `Cannot update missing record 'agreement.market1'`
Selection_268

Create database, not demo data

Why do you add data file https://github.com/OCA/contract/pull/669/files#diff-289fece69d3e767eb82c1fa91fff8a6ae1b74aa76123ef49d0c1aaf49532c9a0R21 ? I saw demo data  in demo folder and data folder, it's same ?

**Comment by patrickrwilson** [comment]
@newtratip Please see new PR #696 that removes that data/demo.xml which shouldn't be there and fixes this installation error. Your review would be appreciated.

Thanks

**Pull request merged**
